wdl-lint/src/v1/preamble_whitespace.rs

```diff
@@ -0,0 +1,192 @@
+    }
+
+    fn explanation(&self) -> &'static str {
+        "The document preamble is defined as anything before the version declaration statement and \
+         the version declaration statement itself. Only comments and whitespace are permitted \
+         before the version declaration. If there are no comments, the version declaration must be \
+         the first line of the document. If there are comments, there must be exactly one blank \
+         line between the last comment and the version declaration."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Spacing, Tag::Style])
+    }
+
+    fn visitor(&self) -> Box<dyn Visitor<State = Diagnostics>> {
+        Box::new(PreambleWhitespaceVisitor::default())
+    }
+}
+
+/// Implements the visitor for the preamble whitespace rule.
+#[derive(Default, Debug)]
+struct PreambleWhitespaceVisitor {
+    /// Whether or not the rule has finished.
+    finished: bool,
+}
+
+impl Visitor for PreambleWhitespaceVisitor {
+    type State = Diagnostics;
+
+    fn version_statement(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        stmt: &VersionStatement,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // We're finished after the version statement
+        self.finished = true;
+
+        // If the previous token is whitespace and its previous token is a comment,
+        // then the whitespace should just be two lines
+        match stmt
+            .syntax()
+            .prev_sibling_or_token()
+            .and_then(SyntaxElement::into_token)
+        {
+            Some(whitespace) if whitespace.kind() == SyntaxKind::Whitespace => {
+                let range = whitespace.text_range();
+                if whitespace
+                    .prev_sibling_or_token()
+                    .map(|s| s.kind() == SyntaxKind::Comment)
+                    .unwrap_or(false)
+                {
+                    // There's a preceding comment, so it is expected that the whitespace is only
+                    // two newlines in a row
+                    let text = whitespace.text();
+                    let mut count = 0;
+                    for (line, start, next_start) in lines_with_offset(text) {
+                        // If this is not a blank line, it's unnecessary whitespace
+                        if !line.is_empty() {
+                            state.add(unnecessary_whitespace(Span::new(
+                                usize::from(range.start()) + start,
+                                line.len(),
+                            )));
+                        }
+
+                        count += 1;
+                        if count == 2 {
+                            // If the previous byte isn't a newline, then we are missing a blank
+                            // line
+                            if text.as_bytes()[next_start - 1] != b'\n' {
```

Comment:
Can you explain this check to me? I'm not following what this `if` statement accomplishes

Reply:
It ensures that there were two actual lines as `lines_with_offsets` will return back a line even if there's no line terminator at the end (i.e. if it can't find a line terminator it returns the remainder of the string as the last "line").

Imagine you had the following:

```wdl
## Hello
  version 1.1
```

The whitespace token would be `\n  ` (a newline with two spaces after it). `lines_with_offsets` returns one blank line and one line with two spaces in it; if we don't check to see that the current line doesn't end with a `\n`, we'd fail to report the missing expected blank line between the last comment and the version statement.

Consequently, we'd report this as two warnings: unnecessary whitespace for the two spaces and a missing blank line between last comment and version statement.

Reply:
That makes sense. Thanks!